src/ipa_pc/mod.rs

```diff
@@ -378,106 +378,126 @@ impl<G: AffineCurve, D: Digest> InnerProductArgPC<G, D> {
-                    .ok_or(Error::MissingEvaluation {
-                        label: label.to_string(),
-                    })?
-            ];
-            v_values.push(v_i);
-
-            points.push(point);
-        }
-
-        // Commitment of the h(X) polynomial
-        let batch_commitment = batch_proof.batch_commitment.clone();
 
-        // lambda
-        let lambda: G::ScalarField = fs_rng.squeeze_128_bits_challenge();
-        let mut cur_challenge = G::ScalarField::one();
+            // (X - x_i)
+            let x_polynomial = Polynomial::from_coefficients_slice(&[
+                -(*point),
+                G::ScalarField::one()
+            ]);
 
-        // Fresh random challenge x
-        fs_rng.absorb(&to_bytes![batch_commitment].unwrap());
-        let point: G::ScalarField = fs_rng.squeeze_128_bits_challenge();
+            // z_i(X) = z(X) / (X - x_i)
+            let z_i_polynomial = &z_polynomial / &x_polynomial;
 
-        let mut computed_batch_v = G::ScalarField::zero();
+            // z_i(x)
+            let z_i_value = z_i_polynomial.evaluate(x_point);
 
-        for ((&v_i, y_i), x_i) in v_values.iter().zip(y_values).zip(points) {
+            // LC(C) = SUM ( lambda^i * z_i(x) * C_i )
+            for (index, &comm_dash_segment) in labeled_commitment.commitment().comm.iter().enumerate() {
+                if index == lc_comm.len() {
+                    lc_comm.push(G::zero());
+                }
+                lc_comm[index] = (lc_comm[index].into_projective()
+                    + &(comm_dash_segment.into_projective()
+                        .mul(&z_i_value)
+                        .mul(&cur_challenge)
+                    )
+                ).into_affine();
+            }
 
-            computed_batch_v = computed_batch_v + &(cur_challenge * &((v_i - &y_i) / &(point - x_i)));
+            lc_value += y_i * cur_challenge * z_i_value;
 
+            // lambda^i
             cur_challenge = cur_challenge * &lambda;
         }
 
-        let mut batch_values = vec![];
-        for labeled_commitment in commitments.iter() {
-            batch_values.push(batch_proof.batch_values[
-                *labels
-                    .get(labeled_commitment.label())
-                    .ok_or(Error::MissingEvaluation {
-                        label: labeled_commitment.label().to_string(),
-                    })?
-                ]
-            );
-        }
+        // z(x)
+        let z_x_value = z_polynomial.evaluate(x_point);
+
+        // LC(C) = SUM ( lambda^i * z_i(x) * C_i ) - z(x) * C
+        lc_comm = lc_comm.iter().enumerate().map(|(index, &lc_comm_segment)| {
+            let h_comm_segment = if batch_proof.h_comm.len() > index {
+                batch_proof.h_comm[index]
+            } else {
+                G::zero()
+            };
+            (lc_comm_segment.into_projective() + &h_comm_segment.mul(-z_x_value)).into_affine()
+        }).collect::<Vec<_>>();
```

Comment:
In order to improve further the large scale batch verifier, let us even work with the following equivalent LC:
Instead of showing that ``LC = SUM ( lambda^i * z_i(x) * C_i ) - z(x) * C`` opens to ``v = SUM (lambda^i)*z_i(x)*y_i``, we show that 
```
LC' = SUM ( lambda^i * z_i(x)/z(x) * C_i ) -  C
``` 
opens to ``v' = 1/z(x) * SUM (lambda^i)*z_i(x)*y_i`` instead.